lib/journaled/kinesis_batch_sender.rb

```diff
@@ -0,0 +1,104 @@
+  #
+  # This class handles:
+  # - Sending events individually to support guaranteed ordering
+  # - Handling failures on a per-event basis
+  # - Classifying errors as transient vs permanent
+  #
+  # Returns structured results for the caller to handle event state management.
+  class KinesisBatchSender
+    FailedEvent = Struct.new(:event, :error_code, :error_message, :transient, keyword_init: true) do
+      def transient?
+        transient
+      end
+
+      def permanent?
+        !transient
+      end
+    end
+
+    TRANSIENT_ERROR_CLASSES = [
+      Aws::Kinesis::Errors::InternalFailure,
+      Aws::Kinesis::Errors::ServiceUnavailable,
+      Aws::Kinesis::Errors::Http503Error,
+      Aws::Kinesis::Errors::ProvisionedThroughputExceededException,
+      Seahorse::Client::NetworkingError,
+    ].freeze
+
+    # Send a batch of database events to Kinesis
+    #
+    # Sends events one at a time to guarantee ordering. Stops on first transient failure.
+    #
+    # @param events [Array<Journaled::Outbox::Event>] Events to send
+    # @return [Hash] Result with:
+    #   - succeeded: Array of successfully sent events
+    #   - failed: Array of FailedEvent structs (only permanent failures)
+    def send_batch(events)
+      result = { succeeded: [], failed: [] }
+
+      events.each do |event|
+        event_result = send_event(event)
+        if event_result.is_a?(FailedEvent)
+          if event_result.transient?
+            emit_transient_failure_metric
+            break
+          else
+            result[:failed] << event_result
+          end
+        else
+          result[:succeeded] << event_result
+        end
+      end
+
+      result
+    end
+
+    private
+
+    # Send a single event to Kinesis
+    #
+    # @param event [Journaled::Outbox::Event] Event to send
+    # @return [Journaled::Outbox::Event, FailedEvent] The event on success, or FailedEvent on failure
+    def send_event(event)
+      # Merge the DB-generated ID into the event data before sending to Kinesis
+      event_data_with_id = event.event_data.merge(id: event.id)
+
+      kinesis_client.put_record(
+        stream_name: event.stream_name,
+        data: event_data_with_id.to_json,
+        partition_key: event.partition_key,
+      )
+
+      event
+    rescue *TRANSIENT_ERROR_CLASSES => e
+      # Event failed with transient error
+      Rails.logger.error("Kinesis event send failed (transient): #{e.class} - #{e.message}")
+      FailedEvent.new(
+        event:,
+        error_code: e.class.to_s,
+        error_message: e.message,
+        transient: true,
+      )
+    rescue StandardError => e
+      # Event failed with permanent error
+      Rails.logger.error("Kinesis event send failed (permanent): #{e.class} - #{e.message}")
+      FailedEvent.new(
+        event:,
+        error_code: e.class.to_s,
+        error_message: e.message,
+        transient: false,
+      )
+    end
```

Comment:
The one catch here is that if we're missing any TRANSIENT_ERROR_CLASSES, this will stop retrying jobs that it should maybe actually retry.

Given that I can basically never trust Ruby to tell me the exhaustive list of exceptions a method might throw, is there value in inverting this and rescuing just the known-permanent ones instead?

My understanding is that the only exception representing "poison pill" payloads is `Aws::Kinesis::Errors::ValidationError`, so maybe we set `transient: false` in that case and then `transient: true` in all other cases (and then monitor for the potential of a different form of "poison pill" payload coming from elsewhere in the stack...)?